providers/implementations/kem/template_kem.c

```diff
@@ -0,0 +1,183 @@
+static OSSL_FUNC_kem_freectx_fn template_freectx;
+static OSSL_FUNC_kem_set_ctx_params_fn template_set_ctx_params;
+
+static void *template_newctx(void *provctx)
+{
+    PROV_TEMPLATE_CTX *ctx = OPENSSL_zalloc(sizeof(*ctx));
+
+    debug_print("newctx called\n");
+    if (ctx == NULL)
+        return NULL;
+    ctx->libctx = PROV_LIBCTX_OF(provctx);
+
+    debug_print("newctx returns %p\n", ctx);
+    return ctx;
+}
+
+static void template_freectx(void *vctx)
+{
+    PROV_TEMPLATE_CTX *ctx = (PROV_TEMPLATE_CTX *)vctx;
+
+    debug_print("freectx %p\n", ctx);
+    OPENSSL_free(ctx);
+}
+
+static int template_init(void *vctx, int operation, void *vkey, void *vauth,
+                         ossl_unused const OSSL_PARAM params[])
+{
+    PROV_TEMPLATE_CTX *ctx = (PROV_TEMPLATE_CTX *)vctx;
+
+    debug_print("init %p / %p\n", ctx, vkey);
+    if (!ossl_prov_is_running())
+        return 0;
+
+    /* check and fill in reference to key */
+    ctx->op = operation;
+    debug_print("init OK\n");
+    return 1;
+}
+
+static int template_encapsulate_init(void *vctx, void *vkey,
+                                     const OSSL_PARAM params[])
+{
+    return template_init(vctx, EVP_PKEY_OP_ENCAPSULATE, vkey, NULL, params);
+}
+
+static int template_decapsulate_init(void *vctx, void *vkey,
+                                     const OSSL_PARAM params[])
+{
+    return template_init(vctx, EVP_PKEY_OP_DECAPSULATE, vkey, NULL, params);
+}
+
+static int template_set_ctx_params(void *vctx, const OSSL_PARAM params[])
+{
+    PROV_TEMPLATE_CTX *ctx = (PROV_TEMPLATE_CTX *)vctx;
+
+    debug_print("set ctx params %p\n", ctx);
+    if (ctx == NULL)
+        return 0;
+    if (params == NULL)
+        return 1;
+
+    debug_print("set ctx params OK\n");
+    return 1;
+}
+
+static const OSSL_PARAM known_settable_template_ctx_params[] = {
+    /* possibly more params */
+    OSSL_PARAM_END
+};
+
+static const OSSL_PARAM *template_settable_ctx_params(ossl_unused void *vctx,
+                                                      ossl_unused void *provctx)
+{
+    return known_settable_template_ctx_params;
+}
+
+static int template_encapsulate(void *vctx, unsigned char *out, size_t *outlen,
+                                unsigned char *secret, size_t *secretlen)
+{
+    debug_print("encaps %p to %p\n", vctx, out);
+    if (outlen != NULL)
+        *outlen = 0; /* replace with real encapsulated data length */
+    if (secretlen != NULL)
+        *secretlen = 0; /* replace with real shared secret length */
+
+    if (out == NULL) {
+        debug_print("encaps outlens set to %d and %d\n", *outlen, *secretlen);
+        return 1;
+    }
+
+    /* check key and perform real KEM operation */
+
+    debug_print("encaps OK\n");
+    return 1;
+}
+
+static int template_decapsulate(void *vctx, unsigned char *out, size_t *outlen,
+                                const unsigned char *in, size_t inlen)
+{
+    debug_print("decaps %p to %p inlen at %d\n", vctx, out, inlen);
```

Comment:
Keep in mind that recently some code was added into rsa_kem to check the following..
 
    if (outlen != NULL && *outlen < nlen) {
        ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_OUTPUT_LENGTH);
        return 0;
    }
    
    So the code you have here needs to be careful not to trash outlen
    
    i.e outlen is actually inoutlen

Reply:
Agreed; all kinds of length integrity checks are sensible to add -- but again, they are mostly algorithm-specific, so can't easily be done in a generic way in a template.

Reply:
comments added.